image-mdraid.c

```diff
@@ -0,0 +1,226 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+
+MDRAID Superblock generator
+This should create valid mdraid superblock for raid1 with 1 device (more devices can be added once mounted).
+It is still very basic, but following seems to be working:
+
+mdadm --examine test.img
+losetup /dev/loop1 test.img
+mdadm --assemble md /dev/loop1
+
+Some docs:
+https://raid.wiki.kernel.org/index.php/RAID_superblock_formats#Sub-versions_of_the_version-1_superblock
+https://docs.huihoo.com/doxygen/linux/kernel/3.7/md__p_8h_source.html
+
+*/
+
+#include <confuse.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <linux/raid/md_p.h>
+
+#include "genimage.h"
+
+#define DATA_OFFSET_SECTORS	(2048)
+#define DATA_OFFSET_BYTES	(DATA_OFFSET_SECTORS*512)
+
+static void random_uuid(__u8 *buf)
+{
+	__u32 r[4];
+	for (int i = 0; i < 4; i++)
+		r[i] = random();
+	memcpy(buf, r, 16);
+}
+
+static unsigned int calc_sb_1_csum(struct mdp_superblock_1 * sb)
+{
+	unsigned int disk_csum, csum;
+	unsigned long long newcsum;
+	int size = sizeof(*sb) + __le32_to_cpu(sb->max_dev)*2;
+	unsigned int *isuper = (unsigned int*)sb;
+
+	/* make sure I can count... (needs include cstddef) */
+	/*
+	if (offsetof(struct mdp_superblock_1,data_offset) != 128 ||
+	    offsetof(struct mdp_superblock_1, utime) != 192 ||
+	    sizeof(struct mdp_superblock_1) != 256) {
+		fprintf(stderr, "WARNING - superblock isn't sized correctly\n");
+	}
+	*/
+
+	disk_csum = sb->sb_csum;
+	sb->sb_csum = 0;
+	newcsum = 0;
+	for (; size>=4; size -= 4 ) {
+		newcsum += __le32_to_cpu(*isuper);
+		isuper++;
+	}
+
+	if (size == 2)
+		newcsum += __le16_to_cpu(*(unsigned short*) isuper);
+
+	csum = (newcsum & 0xffffffff) + (newcsum >> 32);
+	sb->sb_csum = disk_csum;
+	return __cpu_to_le32(csum);
+}
+
+static int mdraid_generate(struct image *image) {
+	struct image *img_in = image->handler_priv;
+
+	int max_devices = 1;
+
+	char *name = cfg_getstr(image->imagesec, "label");
+
+	struct mdp_superblock_1 *sb = xzalloc(sizeof(struct mdp_superblock_1)+max_devices*2);
+
+	/* constant array information - 128 bytes */
+	sb->magic = 0xa92b4efc;		/* MD_SB_MAGIC: 0xa92b4efc - little endian */
+	sb->major_version = 1;	/* 1 */
+	sb->feature_map = 0; //MD_FEATURE_BITMAP_OFFSET;	/* bit 0 set if 'bitmap_offset' is meaningful */ //TODO: internal bitmap bit is ignored, unless there is correct bitmap with BITMAP_MAGIC in place
+	sb->pad0 = 0;		/* always set to 0 when writing */
+
+	random_uuid(sb->set_uuid);	/* user-space generated. U8[16]*/ //TODO: should we allow user to set this?
```

Comment:
Add an option for this please. It should be possible to make image generation reproducible.

Reply:
to be honest. i was trying to avoid complexity of UUID parsing and handling, but now i've noticed there seems to be comprehensive toolkit in util.c that i can use :-)